src/Moq/Protected/ProtectedMock.cs

```diff
@@ -567,5 +800,24 @@ private static IEnumerable<Expression> ToExpressionArgs(MethodInfo method, objec
 				yield return ToExpressionArg(methodParams[i].ParameterType, args[i]);
 			}
 		}
+
+		private static Type[] GetIndexerTypes(PropertyInfo property, bool isSet)
```

Comment:
`GetIndexerTypes` &rarr; `GetIndexerAccessorParameterTypes`

In fact, I'm not even sure this method is needed at all. This method is called exclusively from `IndexerArgsToExpressions`, which in turn is called from two places:

![image](https://user-images.githubusercontent.com/104481/130349852-5571cab1-023b-468c-a093-296edda54f03.png)

Those callers retrieve an accessor method. Why not pass that forward such that you can simply do a `indexerAccessorMethod.GetParameterTypes()`?

Reply:
Removed GetIndexerTypes and IndexerArgsToExpressions.
Replace with GetIndexerMethodCallExpression.

```
		private static Expression<Action<T>> GetIndexerSetterExpression(PropertyInfo property, object[] indexerKeysAndValue)
		{
			var param = Expression.Parameter(typeof(T), "mock");

			return Expression.Lambda<Action<T>>(
				GetIndexerMethodCallExpression(param, property.GetSetMethod(true), indexerKeysAndValue),
				param);
		}

		private static LambdaExpression GetIndexerGetterExpression(PropertyInfo property, object[] indexerKeys)
		{
			var param = Expression.Parameter(typeof(T), "mock");

			return Expression.Lambda(
				GetIndexerMethodCallExpression(param, property.GetGetMethod(true), indexerKeys),
				param
			);
		}

		private static MethodCallExpression GetIndexerMethodCallExpression(ParameterExpression mockParameter,MethodInfo accessorMethod, object[] indexerArguments)
		{
			var types = accessorMethod.GetParameterTypes();
			var expressionArgs = indexerArguments.Select((a, i) =>
			{
				return ToExpressionArg(types[i], indexerArguments[i]);
			}).ToArray();
			return Expression.Call(mockParameter, accessorMethod, expressionArgs);
		}
```

Is there any reason this cannot become a constant ?
`var param = Expression.Parameter(typeof(T), "mock");`